File: clang/ArrayStack/ArrayStack.c
#include "ArrayStack.h"

void AS_CreateStack(ArrayStack** Stack, int Capacity)
{
    // 스택을 자유 저장소에 생성
    (*Stack) = (ArrayStack*)malloc(sizeof(ArrayStack));

    // 입력된 Capacity만큼의 노드를 자유 저장소에 생성
    (*Stack)->Nodes = (Node*)malloc(sizeof(Node)*Capacity);

    // Capacity 및 Top 초기화
    (*Stack)->Capacity = Capacity;
    (*Stack)->Top = -1;
}

void AS_DestroyStack(ArrayStack* Stack)
{
    // 노드를 자유 저장소에서 해제
    free(Stack->Nodes);

    // 스택을 자유 저장소에서 해제
    free(Stack);
}

// 스택이 다 차면 중지
// void AS_Push(ArrayStack* Stack, ElementType Data)
// {   
//     if (AS_IsFull(Stack))
//     {
//         printf("Stack Is Full\n");
//         exit(EXIT_FAILURE);
//     }
    
//     Stack->Top++;
//     Stack->Nodes[Stack->Top].Data = Data;
// }

// 스택의 용량이 모두 소진되었을 때 현재 용량의 30%만큼 더 늘림
void AS_Push(ArrayStack* Stack, ElementType Data)
{   
    if (AS_IsFull(Stack))
    {
        int NewCapacity = Stack->Capacity + (Stack->Capacity * 3)/10;
        if (NewCapacity == Stack->Capacity)
        {
            NewCapacity = Stack->Capacity + 1;
        }

        Node *NewNodes = realloc(Stack->Nodes, NewCapacity * sizeof(Node));

        if (!NewNodes)
        {
            perror("realloc failed");
            exit(EXIT_FAILURE);
        }

        Stack->Nodes = NewNodes;
        Stack->Capacity = NewCapacity;
    }
    
    Stack->Top++;
    Stack->Nodes[Stack->Top].Data = Data;
}

ElementType AS_Pop(ArrayStack* Stack)
{
    int Position = Stack->Top--;
    ElementType val = Stack->Nodes[Position].Data;
    int NewCapacity = (Stack->Capacity * 7)/10;
    if (NewCapacity < 1) {
        NewCapacity = 1;
    }

    if (Position == NewCapacity)
    {
        Node *NewNode = realloc(Stack->Nodes, NewCapacity * sizeof(Node));

        if (NewNode)
        {
            Stack->Nodes = NewNode;
            Stack->Capacity = NewCapacity;
        }
    }

    return val;
}

ElementType AS_Top(ArrayStack* Stack)
{
    return Stack->Nodes[Stack->Top].Data;
}

int AS_GetSize(ArrayStack* Stack)
{
    return Stack->Top+1;
}

int AS_IsEmpty(ArrayStack* Stack)
{
    return (Stack->Top == -1);
}

int AS_IsFull(ArrayStack* Stack)
{   
    return (AS_GetSize(Stack) == Stack->Capacity);
}

File: pylang/ArrayStack/ArrayStack.py
class Node:
    def __init__(self, Data):
        self.Data = Data

class ArrayStack:
    def __init__(self, Capacity):
        self.Capacity = Capacity
        self.Top = -1
        self.Nodes = [None] * Capacity
    
    # def AS_Push(self, Data):
    #     if self.AS_IsFull():
    #         raise Exception("Stack Overflow")
  
    #     self.Top += 1
    #     self.Nodes[self.Top] = Data

    def AS_Push(self, Data):
        if self.AS_IsFull():
            extra = int(self.Capacity * 0.3)
            if extra == 0:
                extra = 1
            self.Nodes += [None] * extra
            self.Capacity += extra
  
        self.Top += 1
        self.Nodes[self.Top] = Data
    
    # def AS_Pop(self):
    #     if self.AS_IsEmpty():
    #         raise Exception("Stack Underflow")
    #     data = self.Nodes[self.Top]
    #     self.Top -= 1
    #     return data

    def AS_Pop(self):
        if self.AS_IsEmpty():
            raise Exception("Stack Underflow")

        Val = self.Nodes[self.Top]
        self.Nodes[self.Top] = None
        self.Top -= 1

        Current_size = self.AS_GetSize()

        if Current_size == int(self.Capacity * 0.7):
            New_Capacity = Current_size
            self.Nodes = self.Nodes[:New_Capacity]
            self.Capacity = New_Capacity

        return Val

    def AS_GetSize(self):
        return self.Top + 1 

    def AS_IsEmpty(self):
        return self.Top == -1
    
    def AS_IsFull(self):
        return self.AS_GetSize() == self.Capacity

# AS = ArrayStack(4)
# AS.AS_Push(1)
# AS.AS_Push(2)
# AS.AS_Push(3)
# AS.AS_Push(4)
# AS.AS_Push(5)

# print(AS.AS_Pop())
# print(AS.AS_Pop())
# print(AS.AS_Pop())
# print(AS.AS_Pop())
# print(AS.AS_Pop())


AS = ArrayStack(15)
for i in range(15):
    AS.AS_Push(i)

print("Before pop:", AS.Capacity)  # 10

# pop 1개 하면 Top = 9 → size = 9 → pass
# ...
# pop 3개 → size = 7 → 줄어듦
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())
print(AS.AS_Pop())



print("After pop:", AS.Capacity)  # 7